crates/provider-keyvalue-nats/src/config.rs

```diff
@@ -0,0 +1,116 @@
+use std::collections::HashMap;
+
+use anyhow::{bail, Result};
+use serde::{Deserialize, Serialize};
+
+const DEFAULT_NATS_URI: &str = "nats://0.0.0.0:4222";
+
+const CONFIG_NATS_URI: &str = "cluster_uri";
+const CONFIG_NATS_JETSTREAM_DOMAIN: &str = "js_domain";
+const CONFIG_NATS_CLIENT_JWT: &str = "client_jwt";
+const CONFIG_NATS_CLIENT_SEED: &str = "client_seed";
+const CONFIG_NATS_TLS_CA: &str = "tls_ca";
+const CONFIG_NATS_TLS_CA_FILE: &str = "tls_ca_file";
+
+/// Configuration for connecting a NATS client.
+/// More options are available if you use the json than variables in the values string map.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct NatsConnectionConfig {
```

Comment:
Just starting this here as a follow up to our previous conversation about the bucket name. So I think the actual NATS bucket name would be set in the config (the one attached at link time). At runtime, I think we follow what other providers do right now and ignore the bucket name (see [here](https://github.com/wasmCloud/wasmCloud/blob/ffbd0310c5e35f9e10b29674b1d0f63473687be7/examples/rust/components/http-keyvalue-counter/src/lib.rs#L22) for an example of that in action). So you'd add a new config item called `BUCKET_NAME` and that would be specified at link time rather than the user needing to know the bucket name in the database or having to configure it. 

Like I mentioned, this isn't perfectly ideal and we should figure it out, but it matches what our other examples are expecting

Reply:
This is a trivial change to make; but it's on very questionable basis, and at the cost of forcing users into using a single bucket (unless they make a change to their link config); and as someone, who's professionally is an _**operator**_ - and as such very well aware of user grievances - I have to ask why and what security or operational benefits do we gain by depriving users of working with different NATS Kv stores?

I'll use my own use case as an example; I have middleware components, which perform operations in behalf of system users; in the case of Kv store, in order to isolate user data (data privacy laws, ease of data management), a separate bucket will be used for each user; however, this change will make that impossible and would force me to have a fork (i.e a custom provider).

I hear you when you say WASI interfaces usually cover the common denominator functions across various backends; but the `open` function is part of the spec; so, I also don't see that being undermined.

So, I ask you to reconsider. However, I respect you as one of the maintainers and project leads; so, if you come back and say I should do it, then I will.

Reply:
So that 100% is something we want to support and that is completely doable as is. Let me have another go at explaining each of the parts at play here, because I don't think I've done a good job explaining:

The "bucket name" argument that is part of the `open` method is essentially a name that maps to a set of config or existing connection provided by the host. It isn't the actual name of a bucket in a database. This is for two distinct reasons. The first is separation of concerns/non-functional requirements – the developer of the code doesn't need to know exactly what the name of a given bucket or database is, just that it is a valid connection. Second is that the interface needs to be portable. A key value database doesn't necessarily have the concept of a bucket and even if it does, those names could vary across implementations.

This is why we have the concept of link names, so you can link a provider to a component multiple times. There is a function for setting which link you'd like to use that is exposed as part of the wasmCloud bus API: https://github.com/wasmCloud/wasmCloud/blob/ffbd0310c5e35f9e10b29674b1d0f63473687be7/wit/bus.wit#L13. So in a wadm manifest, your config would look like this for multiple users:

```yaml
- type: link
  properties:
    name: user1
    target: nats-kv
    namespace: wasi
    package: keyvalue
    interfaces: [store, atomics]
    target_config:
      - name: user1-connection
        properties:
          # This doesn't exist yet in your code but is what I'm suggesting
          bucket: user1-bucket
- type: link
  properties:
    name: user2
    target: nats-kv
    namespace: wasi
    package: keyvalue
    interfaces: [store, atomics]
    target_config:
      - name: user2-connection
        properties:
          bucket: user2-bucket
```

Now with that said, I can come back to what I meant with "we haven't decided what we should use that bucket name for yet." This is because we already had support for doing what you're suggesting via link names. The more I have seen from users like yourself, the more I might backtrack on something we've thought about before by having that bucket name just be an alias for a link name. So in your code, if you're using the config I showed above, you'd do `store::open("user1")` to get a handle to the proper bucket rather than calling `set_link_name` and then `open`. In this case, from the code's perspective, what it is exactly connecting to is transparent, but allows you to indicate that you're trying to use a specific handle associated with some ID.

Hopefully that makes more sense now. If not, I'd love to chat either in the community call or via slack about this